consensus/scp/tests/test_mesh_networks.rs

```diff
@@ -0,0 +1,275 @@
+
+    for _i in 0..10 {
+        let n = test_utils::test_node_id(rng.gen_range(0, num_nodes as u32));
+        for _j in 0..2000 {
+            let value = mock_network::random_str(&mut rng, 10);
+            network.push_value(&n, &value);
+            values.push(value);
+        }
+        sleep(Duration::from_millis(rng.gen_range(0, 50)));
+    }
+
+    // Check that the values got added to the nodes
+    for node_num in 0..num_nodes {
+        let node_id = test_utils::test_node_id(node_num as u32);
+
+        network.wait_for_total_values(&node_id, values.len(), Duration::from_secs(1200));
+
+        assert_eq!(
+            values.iter().cloned().collect::<HashSet<String>>(),
+            network
+                .get_shared_data(&node_id)
+                .get_all_values()
+                .iter()
+                .cloned()
+                .collect::<HashSet<String>>()
+        );
+    }
+
+    // Check all blocks in the ledger are the same
+    let node0_data = network.get_shared_data(&test_utils::test_node_id(0)).ledger;
+    assert!(!node0_data.is_empty());
+
+    for node_num in 0..num_nodes {
+        let node_data = network
+            .get_shared_data(&test_utils::test_node_id(node_num as u32))
+            .ledger;
+        assert_eq!(node0_data.len(), node_data.len());
+
+        for block_num in 0..node0_data.len() {
+            assert_eq!(node0_data.get(block_num), node_data.get(block_num));
+        }
+    }
+
+    // Done
+    log::info!(
+        logger,
+        "mesh_test_helper num_nodes={} k={}: {:?} (avg {} tx/s)",
+        num_nodes,
+        k,
+        start.elapsed(),
+        values.len() as u64 / std::cmp::max(1, start.elapsed().as_secs()),
+    );
+}
+
+#[test_with_logger]
+#[serial]
+fn mesh_1(logger: Logger) {
+    mesh_test_helper(1, 0, logger);
+}
+
+#[test_with_logger]
+#[serial]
+fn mesh_2_k1(logger: Logger) {
+    mesh_test_helper(2, 1, logger);
+}
+
+#[test_with_logger]
+#[serial]
+fn mesh_3_k1(logger: Logger) {
+    mesh_test_helper(3, 1, logger);
+}
+
+#[test_with_logger]
+#[serial]
+fn mesh_3_k2(logger: Logger) {
+    mesh_test_helper(3, 2, logger);
+}
+
+#[test_with_logger]
+#[serial]
+fn mesh_4_k3(logger: Logger) {
+    mesh_test_helper(4, 3, logger);
+}
+
+#[test_with_logger]
+#[serial]
+fn mesh_5_k3(logger: Logger) {
+    mesh_test_helper(5, 3, logger);
+}
+
+#[test_with_logger]
+#[serial]
+fn mesh_5_k4(logger: Logger) {
+    mesh_test_helper(5, 4, logger);
+}
+
+// This is a very slow test :(
+#[test_with_logger]
+#[serial]
+fn mesh_9_k7(logger: Logger) {
+    // Since this test is very slow without --release, we skip it for debug builds.
+    // See https://stackoverflow.com/questions/39204908/how-to-check-release-debug-builds-using-cfg-in-rust/39205417#39205417
+    if cfg!(debug_assertions) {
+        return;
+    }
+
+    mesh_test_helper(9, 7, logger);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Other network topologies
```

Comment:
Why are these here? They're not mesh. If you want to break this into separate files (which honestly I don't personally find particularly useful, but I don't feel strongly about) then this shouldn't be here.

Reply:
The fig2 network from the paper doesn't actually have quorom intersection, which is why this test is marked as ignored. I would be happy to just delete it if there are no objections.

Reply:
(I deleted it.)